src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/TypeNameBuilder.cs

```diff
@@ -0,0 +1,442 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+/*
+    This TypeNameBuilder is modified from CoreCLR's original.
+    It replaces the C++ bits of the implementation with a faithful C# port.
+*/
+
+using System.Collections.Generic;
+using System.Collections;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace System.Reflection.Emit
+{
+    internal class TypeNameBuilder
+    {
+        private enum ParseState
+        {
+            START = 0x0001,
+            NAME = 0x0004,
+            GENARGS = 0x0008,
+            PTRARR = 0x0010,
+            BYREF = 0x0020,
+            ASSEMSPEC = 0x0080,
+            ERROR = 0x0100,
+        }
+
+        private ParseState parseState;
```

Comment:
Nit: The libraries coding convention style is to prefix instance variables with `_`

Reply:
Do you plan to address this?

The coding style is documented at https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/coding-style.md